<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>From Novice to Expert in Bioinformatics</title>
  <meta name="description" content="This is a book for people who wants to learn basic knowledges about Bioinformatics">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="From Novice to Expert in Bioinformatics" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.jpg" />
  <meta property="og:description" content="This is a book for people who wants to learn basic knowledges about Bioinformatics" />
  <meta name="github-repo" content="xie186/Novice2Expert4Bioinformatics" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="From Novice to Expert in Bioinformatics" />
  
  <meta name="twitter:description" content="This is a book for people who wants to learn basic knowledges about Bioinformatics" />
  <meta name="twitter:image" content="images/cover.jpg" />

<meta name="author" content="Shaojun Xie">


<meta name="date" content="2017-11-08">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="arrays-in-perl.html">
<link rel="next" href="control-structure.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">From Novice to Expert in Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#who-should-read-this-book"><i class="fa fa-check"></i>Who should read this book?</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#how-to-read-this-book"><i class="fa fa-check"></i>How to read this book?</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#the-reason-why-i-didnt-write-in-chinese"><i class="fa fa-check"></i>The reason why I didn’t write in Chinese?</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-me"><i class="fa fa-check"></i>About me</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#acknowledgement"><i class="fa fa-check"></i>Acknowledgement</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#disclaimer"><i class="fa fa-check"></i>Disclaimer</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="why-linux.html"><a href="why-linux.html"><i class="fa fa-check"></i><b>1</b> Why Linux?</a><ul>
<li class="chapter" data-level="1.1" data-path="why-linux.html"><a href="why-linux.html#what-is-linux"><i class="fa fa-check"></i><b>1.1</b> What is Linux</a></li>
<li class="chapter" data-level="1.2" data-path="why-linux.html"><a href="why-linux.html#linux-for-bioinformatics"><i class="fa fa-check"></i><b>1.2</b> Linux for bioinformatics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html"><i class="fa fa-check"></i><b>2</b> Connecting to Linux</a><ul>
<li class="chapter" data-level="2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#user-interfaces"><i class="fa fa-check"></i><b>2.1</b> User interfaces</a></li>
<li class="chapter" data-level="2.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#how-to-connect"><i class="fa fa-check"></i><b>2.2</b> How to connect</a><ul>
<li class="chapter" data-level="2.2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#set-up-linux-lab-environment"><i class="fa fa-check"></i><b>2.2.1</b> Set up Linux Lab Environment</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#transferring-files-between-local-computer-and-linux-server"><i class="fa fa-check"></i><b>2.3</b> Transferring files between local computer and Linux server</a><ul>
<li class="chapter" data-level="2.3.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#use-command-line-tools"><i class="fa fa-check"></i><b>2.3.1</b> Use command line tools</a></li>
<li class="chapter" data-level="2.3.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#download-files"><i class="fa fa-check"></i><b>2.3.2</b> Download files</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html"><i class="fa fa-check"></i><b>3</b> File system and commands of Linux</a><ul>
<li class="chapter" data-level="3.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#path"><i class="fa fa-check"></i><b>3.1</b> Path</a><ul>
<li class="chapter" data-level="3.1.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#relative-and-absolute-path"><i class="fa fa-check"></i><b>3.1.1</b> Relative and absolute path</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#surfing-in-linux-file-system"><i class="fa fa-check"></i><b>3.2</b> Surfing in Linux file system</a><ul>
<li class="chapter" data-level="3.2.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-pwd"><i class="fa fa-check"></i><b>3.2.1</b> Command <code>pwd</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#comand-ls"><i class="fa fa-check"></i><b>3.2.2</b> Comand <code>ls</code></a></li>
<li class="chapter" data-level="3.2.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cd"><i class="fa fa-check"></i><b>3.2.3</b> Command <code>cd</code></a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#path-shortcuts"><i class="fa fa-check"></i><b>3.3</b> Path shortcuts</a></li>
<li class="chapter" data-level="3.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#manipulations-of-files-and-directories"><i class="fa fa-check"></i><b>3.4</b> Manipulations of files and directories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cp"><i class="fa fa-check"></i><b>3.4.1</b> Command <code>cp</code></a></li>
<li class="chapter" data-level="3.4.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mv"><i class="fa fa-check"></i><b>3.4.2</b> Command <code>mv</code></a></li>
<li class="chapter" data-level="3.4.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mkdir"><i class="fa fa-check"></i><b>3.4.3</b> Command <code>mkdir</code></a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#viewing-text-files-in-linux"><i class="fa fa-check"></i><b>3.5</b> Viewing text files in Linux</a><ul>
<li class="chapter" data-level="3.5.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cat"><i class="fa fa-check"></i><b>3.5.1</b> Command <code>cat</code></a></li>
<li class="chapter" data-level="3.5.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-more-and-less"><i class="fa fa-check"></i><b>3.5.2</b> Command <code>more</code> and <code>less</code></a></li>
<li class="chapter" data-level="3.5.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-head-and-tail"><i class="fa fa-check"></i><b>3.5.3</b> Command <code>head</code> and <code>tail</code></a></li>
<li class="chapter" data-level="3.5.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#auto-completion"><i class="fa fa-check"></i><b>3.5.4</b> Auto-completion</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#understand-standard-input-and-stardard-output"><i class="fa fa-check"></i><b>3.6</b> Understand standard input and stardard output</a><ul>
<li class="chapter" data-level="3.6.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdin"><i class="fa fa-check"></i><b>3.6.1</b> STDIN</a></li>
<li class="chapter" data-level="3.6.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdout"><i class="fa fa-check"></i><b>3.6.2</b> STDOUT</a></li>
<li class="chapter" data-level="3.6.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stderr"><i class="fa fa-check"></i><b>3.6.3</b> STDERR</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#auto-completion-1"><i class="fa fa-check"></i><b>3.7</b> Auto-completion</a></li>
<li class="chapter" data-level="3.8" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#advanced-topic"><i class="fa fa-check"></i><b>3.8</b> Advanced topic</a><ul>
<li class="chapter" data-level="3.8.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#linux-md5sum-command"><i class="fa fa-check"></i><b>3.8.1</b> Linux <code>md5sum</code> Command</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#section"><i class="fa fa-check"></i><b>3.9</b> </a></li>
<li class="chapter" data-level="3.10" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#check-you-job-status"><i class="fa fa-check"></i><b>3.10</b> Check you job status</a><ul>
<li class="chapter" data-level="3.10.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#list-files-bigger-than-filesize-specified"><i class="fa fa-check"></i><b>3.10.1</b> List files bigger than filesize specified</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html"><i class="fa fa-check"></i><b>4</b> Text editor in Linux</a><ul>
<li class="chapter" data-level="4.1" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#basic-vi-skills"><i class="fa fa-check"></i><b>4.1</b> Basic vi skills</a></li>
<li class="chapter" data-level="4.2" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#create-new-text-file-with-vi"><i class="fa fa-check"></i><b>4.2</b> Create new text file with <code>vi</code></a></li>
<li class="chapter" data-level="4.3" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#an-example-for-using-editor-r"><i class="fa fa-check"></i><b>4.3</b> An example for using editor R</a><ul>
<li class="chapter" data-level="4.3.1" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#shell-scipt"><i class="fa fa-check"></i><b>4.3.1</b> Shell scipt</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#how-to-exit-vim"><i class="fa fa-check"></i><b>4.4</b> How to exit VIM</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html"><i class="fa fa-check"></i><b>5</b> Install Bioinformatics software in Linux</a><ul>
<li class="chapter" data-level="5.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installation-from-source-code"><i class="fa fa-check"></i><b>5.1</b> Installation from source code</a><ul>
<li class="chapter" data-level="5.1.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa"><i class="fa fa-check"></i><b>5.1.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="5.1.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-samtools"><i class="fa fa-check"></i><b>5.1.2</b> Install <code>samtools</code></a></li>
<li class="chapter" data-level="5.1.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files"><i class="fa fa-check"></i><b>5.1.3</b> Align reads to genome using <code>bwa</code> and store the alignment results in SAM/BAM files</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installing-a-precompiled-binary-executable"><i class="fa fa-check"></i><b>5.2</b> Installing a precompiled binary (executable)</a><ul>
<li class="chapter" data-level="5.2.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa-1"><i class="fa fa-check"></i><b>5.2.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="5.2.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#achiving-and-compressing-files"><i class="fa fa-check"></i><b>5.2.2</b> Achiving and compressing files</a></li>
<li class="chapter" data-level="5.2.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#section-1"><i class="fa fa-check"></i><b>5.2.3</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="first-perl-program.html"><a href="first-perl-program.html"><i class="fa fa-check"></i><b>6</b> First Perl Program</a><ul>
<li class="chapter" data-level="6.1" data-path="first-perl-program.html"><a href="first-perl-program.html#first-program"><i class="fa fa-check"></i><b>6.1</b> First Program</a><ul>
<li class="chapter" data-level="6.1.1" data-path="first-perl-program.html"><a href="first-perl-program.html#how-to-read-extract-a-set-of-sequences-from-the-reference-genome"><i class="fa fa-check"></i><b>6.1.1</b> How to read extract a set of sequences from the reference genome</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html"><i class="fa fa-check"></i><b>7</b> Arrays in Perl</a><ul>
<li class="chapter" data-level="7.1" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#array-index"><i class="fa fa-check"></i><b>7.1</b> Array index</a></li>
<li class="chapter" data-level="7.2" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#sort-arrays-in-perl"><i class="fa fa-check"></i><b>7.2</b> Sort Arrays in Perl</a><ul>
<li class="chapter" data-level="7.2.1" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#sort-alphebetically"><i class="fa fa-check"></i><b>7.2.1</b> Sort alphebetically</a></li>
<li class="chapter" data-level="7.2.2" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#sort-numerically"><i class="fa fa-check"></i><b>7.2.2</b> Sort numerically</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#first-program-1"><i class="fa fa-check"></i><b>7.3</b> First Program</a></li>
<li class="chapter" data-level="7.4" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#vector"><i class="fa fa-check"></i><b>7.4</b> Vector</a></li>
<li class="chapter" data-level="7.5" data-path="arrays-in-perl.html"><a href="arrays-in-perl.html#dataframe"><i class="fa fa-check"></i><b>7.5</b> Dataframe</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="varible-and-arithmetic-operators.html"><a href="varible-and-arithmetic-operators.html"><i class="fa fa-check"></i><b>8</b> Varible and arithmetic operators</a><ul>
<li class="chapter" data-level="8.1" data-path="varible-and-arithmetic-operators.html"><a href="varible-and-arithmetic-operators.html#varaible"><i class="fa fa-check"></i><b>8.1</b> Varaible</a><ul>
<li class="chapter" data-level="8.1.1" data-path="varible-and-arithmetic-operators.html"><a href="varible-and-arithmetic-operators.html#scalar-variable"><i class="fa fa-check"></i><b>8.1.1</b> scalar variable</a></li>
<li class="chapter" data-level="8.1.2" data-path="varible-and-arithmetic-operators.html"><a href="varible-and-arithmetic-operators.html#array"><i class="fa fa-check"></i><b>8.1.2</b> Array</a></li>
<li class="chapter" data-level="8.1.3" data-path="varible-and-arithmetic-operators.html"><a href="varible-and-arithmetic-operators.html#hash"><i class="fa fa-check"></i><b>8.1.3</b> Hash</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="control-structure.html"><a href="control-structure.html"><i class="fa fa-check"></i><b>9</b> Control structure</a><ul>
<li class="chapter" data-level="9.1" data-path="control-structure.html"><a href="control-structure.html#while-loop"><i class="fa fa-check"></i><b>9.1</b> <code>while</code> loop</a></li>
<li class="chapter" data-level="9.2" data-path="control-structure.html"><a href="control-structure.html#for-loop"><i class="fa fa-check"></i><b>9.2</b> <code>for</code> loop</a></li>
<li class="chapter" data-level="9.3" data-path="control-structure.html"><a href="control-structure.html#foreach-loop"><i class="fa fa-check"></i><b>9.3</b> <code>foreach</code> loop</a></li>
<li class="chapter" data-level="9.4" data-path="control-structure.html"><a href="control-structure.html#statement-if-else"><i class="fa fa-check"></i><b>9.4</b> Statement if-else</a></li>
<li class="chapter" data-level="9.5" data-path="control-structure.html"><a href="control-structure.html#operator-last-and-next"><i class="fa fa-check"></i><b>9.5</b> Operator <code>last</code> and <code>next</code></a></li>
<li class="chapter" data-level="9.6" data-path="control-structure.html"><a href="control-structure.html#operator-redo"><i class="fa fa-check"></i><b>9.6</b> Operator <code>redo</code></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="string-manipulation.html"><a href="string-manipulation.html"><i class="fa fa-check"></i><b>10</b> String manipulation</a><ul>
<li class="chapter" data-level="10.1" data-path="string-manipulation.html"><a href="string-manipulation.html#string-concatenation"><i class="fa fa-check"></i><b>10.1</b> String concatenation</a></li>
<li class="chapter" data-level="10.2" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-extraction"><i class="fa fa-check"></i><b>10.2</b> Substring extraction</a></li>
<li class="chapter" data-level="10.3" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-search"><i class="fa fa-check"></i><b>10.3</b> Substring search</a></li>
<li class="chapter" data-level="10.4" data-path="string-manipulation.html"><a href="string-manipulation.html#split-string"><i class="fa fa-check"></i><b>10.4</b> Split String</a></li>
<li class="chapter" data-level="10.5" data-path="string-manipulation.html"><a href="string-manipulation.html#regular-expression"><i class="fa fa-check"></i><b>10.5</b> Regular expression</a><ul>
<li class="chapter" data-level="10.5.1" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-matching"><i class="fa fa-check"></i><b>10.5.1</b> Pattern matching</a></li>
<li class="chapter" data-level="10.5.2" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-substitution"><i class="fa fa-check"></i><b>10.5.2</b> Pattern substitution</a></li>
<li class="chapter" data-level="10.5.3" data-path="string-manipulation.html"><a href="string-manipulation.html#modifiers-to-pattern-matching-and-substitution"><i class="fa fa-check"></i><b>10.5.3</b> Modifiers to pattern matching and substitution</a></li>
<li class="chapter" data-level="10.5.4" data-path="string-manipulation.html"><a href="string-manipulation.html#greedy-or-non-greedy"><i class="fa fa-check"></i><b>10.5.4</b> Greedy or non-greedy</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html"><i class="fa fa-check"></i><b>11</b> Input and output in Perl</a><ul>
<li class="chapter" data-level="11.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input"><i class="fa fa-check"></i><b>11.1</b> Input</a><ul>
<li class="chapter" data-level="11.1.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#standard-input"><i class="fa fa-check"></i><b>11.1.1</b> Standard input</a></li>
<li class="chapter" data-level="11.1.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input-from-a-file-on-the-disk"><i class="fa fa-check"></i><b>11.1.2</b> Input from a file on the disk</a></li>
<li class="chapter" data-level="11.1.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#special-variable-_"><i class="fa fa-check"></i><b>11.1.3</b> Special variable <code>$_</code></a></li>
<li class="chapter" data-level="11.1.4" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#the-input-record-separator"><i class="fa fa-check"></i><b>11.1.4</b> The input record separator: <code>$/</code></a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#output-to-a-file-on-the-disk"><i class="fa fa-check"></i><b>11.2</b> Output to a file on the disk</a></li>
<li class="chapter" data-level="11.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#arguments-from-command-line"><i class="fa fa-check"></i><b>11.3</b> Arguments from command line</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="perl-modules.html"><a href="perl-modules.html"><i class="fa fa-check"></i><b>12</b> Perl modules</a><ul>
<li class="chapter" data-level="12.1" data-path="perl-modules.html"><a href="perl-modules.html#what-is-a-perl-module"><i class="fa fa-check"></i><b>12.1</b> What is a Perl module</a></li>
<li class="chapter" data-level="12.2" data-path="perl-modules.html"><a href="perl-modules.html#how-to-install-a-perl-module"><i class="fa fa-check"></i><b>12.2</b> How to install a Perl module</a></li>
<li class="chapter" data-level="12.3" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-a-perl-module"><i class="fa fa-check"></i><b>12.3</b> How to use a Perl module</a></li>
<li class="chapter" data-level="12.4" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-bioperl-module"><i class="fa fa-check"></i><b>12.4</b> How to use BioPerl module</a></li>
<li class="chapter" data-level="12.5" data-path="perl-modules.html"><a href="perl-modules.html#how-to-write-a-module"><i class="fa fa-check"></i><b>12.5</b> How to write a module</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html"><i class="fa fa-check"></i><b>13</b> Producing simple graphs</a><ul>
<li class="chapter" data-level="13.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#producing-simple-graphs-using-r"><i class="fa fa-check"></i><b>13.1</b> Producing simple graphs using R</a><ul>
<li class="chapter" data-level="13.1.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#line-charts"><i class="fa fa-check"></i><b>13.1.1</b> Line Charts</a></li>
<li class="chapter" data-level="13.1.2" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#bar-charts"><i class="fa fa-check"></i><b>13.1.2</b> Bar Charts</a></li>
<li class="chapter" data-level="13.1.3" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#histograms"><i class="fa fa-check"></i><b>13.1.3</b> Histograms</a></li>
<li class="chapter" data-level="13.1.4" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#pie-charts"><i class="fa fa-check"></i><b>13.1.4</b> Pie Charts</a></li>
<li class="chapter" data-level="13.1.5" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#dotcharts"><i class="fa fa-check"></i><b>13.1.5</b> Dotcharts</a></li>
<li class="chapter" data-level="13.1.6" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#misc"><i class="fa fa-check"></i><b>13.1.6</b> Misc</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html"><i class="fa fa-check"></i><b>14</b> Preparation of figures for manuscript</a><ul>
<li class="chapter" data-level="14.1" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html#section-2"><i class="fa fa-check"></i><b>14.1</b> </a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="start-a-project.html"><a href="start-a-project.html"><i class="fa fa-check"></i><b>15</b> Start a project</a><ul>
<li class="chapter" data-level="15.1" data-path="start-a-project.html"><a href="start-a-project.html#experimetal-design"><i class="fa fa-check"></i><b>15.1</b> Experimetal design</a><ul>
<li class="chapter" data-level="15.1.1" data-path="start-a-project.html"><a href="start-a-project.html#how-many-biological-replicates"><i class="fa fa-check"></i><b>15.1.1</b> How many biological replicates</a></li>
<li class="chapter" data-level="15.1.2" data-path="start-a-project.html"><a href="start-a-project.html#how-big"><i class="fa fa-check"></i><b>15.1.2</b> How big</a></li>
<li class="chapter" data-level="15.1.3" data-path="start-a-project.html"><a href="start-a-project.html#how-much-data-to-generate"><i class="fa fa-check"></i><b>15.1.3</b> How much data to generate</a></li>
<li class="chapter" data-level="15.1.4" data-path="start-a-project.html"><a href="start-a-project.html#section-3"><i class="fa fa-check"></i><b>15.1.4</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="introduction-of-ngs.html"><a href="introduction-of-ngs.html"><i class="fa fa-check"></i><b>16</b> Introduction of NGS</a><ul>
<li class="chapter" data-level="16.1" data-path="introduction-of-ngs.html"><a href="introduction-of-ngs.html#what-is-ngs"><i class="fa fa-check"></i><b>16.1</b> What is NGS</a></li>
<li class="chapter" data-level="16.2" data-path="introduction-of-ngs.html"><a href="introduction-of-ngs.html#application-of-ngs"><i class="fa fa-check"></i><b>16.2</b> Application of NGS</a></li>
<li class="chapter" data-level="16.3" data-path="introduction-of-ngs.html"><a href="introduction-of-ngs.html#section-4"><i class="fa fa-check"></i><b>16.3</b> </a></li>
<li class="chapter" data-level="16.4" data-path="introduction-of-ngs.html"><a href="introduction-of-ngs.html#usefull-links"><i class="fa fa-check"></i><b>16.4</b> Usefull links:</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="practical-perl-program.html"><a href="practical-perl-program.html"><i class="fa fa-check"></i><b>17</b> Practical Perl program</a><ul>
<li class="chapter" data-level="17.1" data-path="practical-perl-program.html"><a href="practical-perl-program.html#extract-sequences-from-genomic-dna-for-some-specific-regions"><i class="fa fa-check"></i><b>17.1</b> Extract sequences from genomic DNA for some specific regions</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html"><i class="fa fa-check"></i><b>18</b> String manipulation</a><ul>
<li class="chapter" data-level="18.1" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#string-concatenation-1"><i class="fa fa-check"></i><b>18.1</b> String concatenation</a></li>
<li class="chapter" data-level="18.2" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#substring-extraction-1"><i class="fa fa-check"></i><b>18.2</b> Substring extraction</a></li>
<li class="chapter" data-level="18.3" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#substring-search-1"><i class="fa fa-check"></i><b>18.3</b> Substring search</a><ul>
<li class="chapter" data-level="18.3.1" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#index"><i class="fa fa-check"></i><b>18.3.1</b> index</a></li>
<li class="chapter" data-level="18.3.2" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#rindex"><i class="fa fa-check"></i><b>18.3.2</b> rindex</a></li>
</ul></li>
<li class="chapter" data-level="18.4" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#split-string-1"><i class="fa fa-check"></i><b>18.4</b> Split String</a><ul>
<li class="chapter" data-level="18.4.1" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#split"><i class="fa fa-check"></i><b>18.4.1</b> split</a></li>
<li class="chapter" data-level="18.4.2" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#join"><i class="fa fa-check"></i><b>18.4.2</b> join</a></li>
<li class="chapter" data-level="18.4.3" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#tr"><i class="fa fa-check"></i><b>18.4.3</b> tr</a></li>
<li class="chapter" data-level="18.4.4" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#reverse"><i class="fa fa-check"></i><b>18.4.4</b> reverse</a></li>
<li class="chapter" data-level="18.4.5" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#markdown"><i class="fa fa-check"></i><b>18.4.5</b> markdown</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#regular-expression-1"><i class="fa fa-check"></i><b>18.5</b> Regular expression</a><ul>
<li class="chapter" data-level="18.5.1" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#pattern-matching-1"><i class="fa fa-check"></i><b>18.5.1</b> Pattern matching</a></li>
<li class="chapter" data-level="18.5.2" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#pattern-substitution-1"><i class="fa fa-check"></i><b>18.5.2</b> Pattern substitution</a></li>
<li class="chapter" data-level="18.5.3" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#modifiers-to-pattern-matching-and-substitution-1"><i class="fa fa-check"></i><b>18.5.3</b> Modifiers to pattern matching and substitution</a></li>
<li class="chapter" data-level="18.5.4" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#greedy-or-non-greedy-1"><i class="fa fa-check"></i><b>18.5.4</b> Greedy or non-greedy</a></li>
<li class="chapter" data-level="18.5.5" data-path="string-manipulation-1.html"><a href="string-manipulation-1.html#section-5"><i class="fa fa-check"></i><b>18.5.5</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html"><i class="fa fa-check"></i><b>19</b> Heatmap Tutorial</a><ul>
<li class="chapter" data-level="19.1" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#install-pheatmap-package"><i class="fa fa-check"></i><b>19.1</b> Install <code>pheatmap</code> package</a></li>
<li class="chapter" data-level="19.2" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#draw-a-heatmap-for-gene-expression-of-rna-seq-data"><i class="fa fa-check"></i><b>19.2</b> Draw a heatmap for gene expression of RNA-seq data</a></li>
<li class="chapter" data-level="19.3" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#add-the-annotation"><i class="fa fa-check"></i><b>19.3</b> Add the annotation</a></li>
<li class="chapter" data-level="19.4" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section-6"><i class="fa fa-check"></i><b>19.4</b> </a></li>
<li class="chapter" data-level="19.5" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#transfrom-the-data"><i class="fa fa-check"></i><b>19.5</b> Transfrom the data</a></li>
<li class="chapter" data-level="19.6" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-add-annotations"><i class="fa fa-check"></i><b>19.6</b> How to add annotations</a></li>
<li class="chapter" data-level="19.7" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-cut-the-trees"><i class="fa fa-check"></i><b>19.7</b> How to cut the trees</a></li>
<li class="chapter" data-level="19.8" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-get-the-cluster-information-from-the-heatmap"><i class="fa fa-check"></i><b>19.8</b> How to get the cluster information from the heatmap</a></li>
<li class="chapter" data-level="19.9" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section-7"><i class="fa fa-check"></i><b>19.9</b> </a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="data-analysis-of-rna-seq.html"><a href="data-analysis-of-rna-seq.html"><i class="fa fa-check"></i><b>20</b> Data analysis of RNA-seq</a></li>
<li class="chapter" data-level="21" data-path="bs-seq.html"><a href="bs-seq.html"><i class="fa fa-check"></i><b>21</b> BS-seq</a><ul>
<li class="chapter" data-level="21.1" data-path="bs-seq.html"><a href="bs-seq.html#section-8"><i class="fa fa-check"></i><b>21.1</b> </a></li>
<li class="chapter" data-level="21.2" data-path="bs-seq.html"><a href="bs-seq.html#what-is-bs-seq"><i class="fa fa-check"></i><b>21.2</b> What is BS-seq</a></li>
<li class="chapter" data-level="21.3" data-path="bs-seq.html"><a href="bs-seq.html#section-9"><i class="fa fa-check"></i><b>21.3</b> </a></li>
<li class="chapter" data-level="21.4" data-path="bs-seq.html"><a href="bs-seq.html#how-to-analyze-bs-seq-data"><i class="fa fa-check"></i><b>21.4</b> How to analyze BS-seq data</a><ul>
<li class="chapter" data-level="21.4.1" data-path="bs-seq.html"><a href="bs-seq.html#section-10"><i class="fa fa-check"></i><b>21.4.1</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="22" data-path="shell.html"><a href="shell.html"><i class="fa fa-check"></i><b>22</b> Shell</a></li>
<li class="chapter" data-level="23" data-path="capstone-project.html"><a href="capstone-project.html"><i class="fa fa-check"></i><b>23</b> Capstone project:</a><ul>
<li class="chapter" data-level="23.1" data-path="capstone-project.html"><a href="capstone-project.html#introduction"><i class="fa fa-check"></i><b>23.1</b> Introduction</a></li>
<li class="chapter" data-level="23.2" data-path="capstone-project.html"><a href="capstone-project.html#method"><i class="fa fa-check"></i><b>23.2</b> Method</a></li>
<li class="chapter" data-level="23.3" data-path="capstone-project.html"><a href="capstone-project.html#pipelines"><i class="fa fa-check"></i><b>23.3</b> Pipelines</a></li>
<li class="chapter" data-level="23.4" data-path="capstone-project.html"><a href="capstone-project.html#section-11"><i class="fa fa-check"></i><b>23.4</b> </a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html"><i class="fa fa-check"></i><b>24</b> Good resouces to learn Bioinformatics</a><ul>
<li class="chapter" data-level="24.1" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html#references"><i class="fa fa-check"></i><b>24.1</b> References:</a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="python.html"><a href="python.html"><i class="fa fa-check"></i><b>25</b> Python</a><ul>
<li class="chapter" data-level="25.1" data-path="python.html"><a href="python.html#os"><i class="fa fa-check"></i><b>25.1</b> os</a></li>
<li class="chapter" data-level="25.2" data-path="python.html"><a href="python.html#section-12"><i class="fa fa-check"></i><b>25.2</b> </a><ul>
<li class="chapter" data-level="25.2.1" data-path="python.html"><a href="python.html#how-to-write-python-script-to-a-command-line-program"><i class="fa fa-check"></i><b>25.2.1</b> How to write python script to a command line program</a></li>
<li class="chapter" data-level="25.2.2" data-path="python.html"><a href="python.html#python-module-pysam"><i class="fa fa-check"></i><b>25.2.2</b> Python module: pysam</a></li>
<li class="chapter" data-level="25.2.3" data-path="python.html"><a href="python.html#python-module-pybedtools"><i class="fa fa-check"></i><b>25.2.3</b> Python module: pybedtools</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="26" data-path="r-introduction.html"><a href="r-introduction.html"><i class="fa fa-check"></i><b>26</b> R introduction</a><ul>
<li class="chapter" data-level="26.1" data-path="r-introduction.html"><a href="r-introduction.html#basic-r-function"><i class="fa fa-check"></i><b>26.1</b> Basic R function</a></li>
<li class="chapter" data-level="26.2" data-path="r-introduction.html"><a href="r-introduction.html#producing-simple-graphs-with-r"><i class="fa fa-check"></i><b>26.2</b> Producing Simple Graphs with R</a><ul>
<li class="chapter" data-level="26.2.1" data-path="r-introduction.html"><a href="r-introduction.html#line-charts-1"><i class="fa fa-check"></i><b>26.2.1</b> Line Charts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">From Novice to Expert in Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="varible-and-arithmetic-operators" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Varible and arithmetic operators</h1>
<div id="varaible" class="section level2">
<h2><span class="header-section-number">8.1</span> Varaible</h2>
<p>Perl provides three kinds of variables: scalars, arrays, and associative arrays. The initial character of the name identifies the particular type of variable and, hence, its functionality.</p>
<div id="scalar-variable" class="section level3">
<h3><span class="header-section-number">8.1.1</span> scalar variable</h3>
<p>either a number or string; Perl does not differentiate between the two, nor does it differentiate between integers and reals. In order to tell the computer what to print, we need to use variables. In Perl, the name of a variable starts with the dollar sign <code>$</code>. You can assign either a number or a string to it.</p>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="co">#assign two strings to two variables</span>
<span class="kw">my</span> <span class="dt">$dna_seq1</span> = <span class="kw">&quot;</span><span class="st">ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna_seq2</span> = <span class="kw">&quot;</span><span class="st">TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna_seq1</span>; 
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna_seq2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna_seq1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna_seq2</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calculate the total length of two DNA sequences</span>
<span class="kw">my</span> <span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_length2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co"># calculate the average length of two DNA sequences. </span>
<span class="kw">my</span> <span class="dt">$average_length</span> = (<span class="dt">$dna_length1</span> + <span class="dt">$dna_length2</span>) / <span class="dv">2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The average length of two DNA sequences is </span><span class="dt">$average_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calulate the differences between the two DNA sequences</span>
<span class="kw">my</span> <span class="dt">$diff_length</span> = <span class="dt">$dna_length2</span> - <span class="dt">$dna_length1</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length difference of two DNA sequences is </span><span class="kw">&quot;</span>, <span class="dt">$diff_length</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#Imaging the CC content of first DNA sequence is 0.5. </span>
<span class="co">#How many CC do we have in first DNA?</span>
<span class="kw">my</span> <span class="dt">$cg_content</span> = <span class="fl">0.5</span>;
<span class="kw">my</span> <span class="dt">$cg_number</span> = <span class="dt">$dna_length1</span> <span class="kw">*</span> <span class="fl">0.5</span>;

<span class="co">#Imaging we have a 10 bps DNA sequnces, </span>
<span class="co">#how many possible DNA sequences do we have?</span>
<span class="kw">my</span> <span class="dt">$dna_nucleotide</span> = <span class="kw">&quot;</span><span class="st">ATCG</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna_nucleo_number</span> = <span class="fu">length</span> <span class="dt">$dna_nucleotide</span>;
<span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;
<span class="kw">my</span> <span class="dt">$possible_number</span> = <span class="dt">$dna_nucleo_number</span> *<span class="kw">*</span> <span class="dt">$dna_length</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">We have </span><span class="dt">$possible_number</span><span class="st"> possibilities.</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre></div>
<p>Here are the explanations of this script.</p>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="co"># assign two DNA sequences to two variables</span>
<span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;</code></pre></div>
<p>The first line here is a comment. In Perl, <code>#</code> (pound sign) is the comment character. The comments can be used to document the program and improve the readability. During the execution, the comments will be ignored.</p>

<div class="rmdtip">
<p><strong>Mutilple lines comments</strong></p>
<p>The quick-and-dirty method only works well when you don’t plan to leave the commented code in the source. If a Pod parser comes along, your multiline comment is going to show up in the Pod translation. A better way hides it from Pod parsers as well.</p>
The =begin directive can mark a section for a particular purpose. If the Pod parser doesn’t want to handle it, it just ignores it. Label the comments with comment. End the comment using =end with the same label. You still need the =cut to go back to Perl code from the Pod comment:
</div>

<p>The following two</p>
<div id="use-strict-user-warnings-my" class="section level4">
<h4><span class="header-section-number">8.1.1.1</span> use strict; user warnings, my</h4>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>

<span class="co">#assign 2 numbers to 2 variable</span>
<span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;
<span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calculate the total length of two DNA sequences</span>
<span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre></div>
<p>In the script above, <code>$dna_lenght2</code> is a typo. If you run this script, it will give you the output without error message, although it’s not the right output.</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/var_assign_no_strict_warnings.pl</code></pre></div>
<pre><code>First DNA: CTCGACCAGGACGATGAATGGGCGATGAAAATCT; Length: 34
Second DNA: CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT; Length: 48
The total length of two DNA sequences is 34 </code></pre>
<div class="sourceCode"><pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="co">#assign 2 numbers to 2 variable</span>
<span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calculate the total length of two DNA sequences</span>
<span class="kw">my</span> <span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre></div>
<p>If we add <code>use strict</code> and <code>use warnings</code>,</p>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name at code_perl/var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted due to compilation errors.</code></pre>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name 
at code_perl/var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted
 due to compilation errors.</code></pre>
</div>
</div>
<div id="array" class="section level3">
<h3><span class="header-section-number">8.1.2</span> Array</h3>
</div>
<div id="hash" class="section level3">
<h3><span class="header-section-number">8.1.3</span> Hash</h3>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="arrays-in-perl.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="control-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/08-PerlVariableOperator.Rmd",
"text": "Edit"
},
"download": ["bioinfBookXIE186.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
