<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 10 Varible in Perl | All things considered for Bioinformatics</title>
  <meta name="description" content="This is a book for people who wants to learn Bioinformatics from scratch">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 10 Varible in Perl | All things considered for Bioinformatics" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.jpg" />
  <meta property="og:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="github-repo" content="xie186/Novice2Expert4Bioinformatics" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Varible in Perl | All things considered for Bioinformatics" />
  
  <meta name="twitter:description" content="This is a book for people who wants to learn Bioinformatics from scratch" />
  <meta name="twitter:image" content="images/cover.jpg" />

<meta name="author" content="Shaojun Xie (解少俊)">


<meta name="date" content="2019-05-01">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="first-perl-program.html">
<link rel="next" href="control-structure.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">From Novice to Expert in Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="part"><span><b>I Linux</b></span></li>
<li class="chapter" data-level="1" data-path="why-linux.html"><a href="why-linux.html"><i class="fa fa-check"></i><b>1</b> Why Linux?</a><ul>
<li class="chapter" data-level="1.1" data-path="why-linux.html"><a href="why-linux.html#what-is-linux"><i class="fa fa-check"></i><b>1.1</b> What is Linux</a></li>
<li class="chapter" data-level="1.2" data-path="why-linux.html"><a href="why-linux.html#linux-for-bioinformatics"><i class="fa fa-check"></i><b>1.2</b> Linux for bioinformatics</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html"><i class="fa fa-check"></i><b>2</b> Connecting to Linux</a><ul>
<li class="chapter" data-level="2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#user-interfaces"><i class="fa fa-check"></i><b>2.1</b> User interfaces</a></li>
<li class="chapter" data-level="2.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#how-to-connect"><i class="fa fa-check"></i><b>2.2</b> How to connect</a><ul>
<li class="chapter" data-level="2.2.1" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#digitalocean-droplet"><i class="fa fa-check"></i><b>2.2.1</b> DigitalOcean droplet</a></li>
<li class="chapter" data-level="2.2.2" data-path="connecting-to-linux.html"><a href="connecting-to-linux.html#aws-free-tier-offering"><i class="fa fa-check"></i><b>2.2.2</b> AWS Free Tier offering</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html"><i class="fa fa-check"></i><b>3</b> File system and commands of Linux</a><ul>
<li class="chapter" data-level="3.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#path"><i class="fa fa-check"></i><b>3.1</b> Path</a><ul>
<li class="chapter" data-level="3.1.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#relative-and-absolute-path"><i class="fa fa-check"></i><b>3.1.1</b> Relative and absolute path</a></li>
<li class="chapter" data-level="3.1.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#comand-ls"><i class="fa fa-check"></i><b>3.1.2</b> Comand <code>ls</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cd"><i class="fa fa-check"></i><b>3.1.3</b> Command <code>cd</code></a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#manipulations-of-files-and-directories"><i class="fa fa-check"></i><b>3.2</b> Manipulations of files and directories</a><ul>
<li class="chapter" data-level="3.2.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cp"><i class="fa fa-check"></i><b>3.2.1</b> Command <code>cp</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mv"><i class="fa fa-check"></i><b>3.2.2</b> Command <code>mv</code></a></li>
<li class="chapter" data-level="3.2.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-mkdir"><i class="fa fa-check"></i><b>3.2.3</b> Command <code>mkdir</code></a></li>
<li class="chapter" data-level="3.2.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-rm"><i class="fa fa-check"></i><b>3.2.4</b> Command ‘rm’</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#viewing-text-files-in-linux"><i class="fa fa-check"></i><b>3.3</b> Viewing text files in Linux</a><ul>
<li class="chapter" data-level="3.3.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-cat"><i class="fa fa-check"></i><b>3.3.1</b> Command <code>cat</code></a></li>
<li class="chapter" data-level="3.3.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-more-and-less"><i class="fa fa-check"></i><b>3.3.2</b> Command <code>more</code> and <code>less</code></a></li>
<li class="chapter" data-level="3.3.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#command-head-and-tail"><i class="fa fa-check"></i><b>3.3.3</b> Command <code>head</code> and <code>tail</code></a></li>
<li class="chapter" data-level="3.3.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#auto-completion"><i class="fa fa-check"></i><b>3.3.4</b> Auto-completion</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#understand-standard-input-and-stardard-output"><i class="fa fa-check"></i><b>3.4</b> Understand standard input and stardard output</a><ul>
<li class="chapter" data-level="3.4.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdin"><i class="fa fa-check"></i><b>3.4.1</b> STDIN</a></li>
<li class="chapter" data-level="3.4.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stdout"><i class="fa fa-check"></i><b>3.4.2</b> STDOUT</a></li>
<li class="chapter" data-level="3.4.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#stderr"><i class="fa fa-check"></i><b>3.4.3</b> STDERR</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#find-disk-usage-of-files-and-directories"><i class="fa fa-check"></i><b>3.5</b> Find Disk Usage of Files and Directories</a></li>
<li class="chapter" data-level="3.6" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#advanced-topic"><i class="fa fa-check"></i><b>3.6</b> Advanced topic</a><ul>
<li class="chapter" data-level="3.6.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#linux-md5sum-command"><i class="fa fa-check"></i><b>3.6.1</b> Linux <code>md5sum</code> Command</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#file-filtering"><i class="fa fa-check"></i><b>3.7</b> File Filtering</a><ul>
<li class="chapter" data-level="3.7.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#column-filtering"><i class="fa fa-check"></i><b>3.7.1</b> Column filtering</a></li>
<li class="chapter" data-level="3.7.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#cut"><i class="fa fa-check"></i><b>3.7.2</b> <code>cut</code></a></li>
<li class="chapter" data-level="3.7.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#row-filtering"><i class="fa fa-check"></i><b>3.7.3</b> Row filtering</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#finding-things"><i class="fa fa-check"></i><b>3.8</b> Finding Things</a><ul>
<li class="chapter" data-level="3.8.1" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#find-files-with-pattern-matching"><i class="fa fa-check"></i><b>3.8.1</b> Find files with pattern matching</a></li>
<li class="chapter" data-level="3.8.2" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#count-file-numbers-in-a-folder-and-its-subdirectories"><i class="fa fa-check"></i><b>3.8.2</b> Count file numbers in a folder and its subdirectories</a></li>
<li class="chapter" data-level="3.8.3" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#list-files-bigger-than-filesize-specified"><i class="fa fa-check"></i><b>3.8.3</b> List files bigger than filesize specified</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="file-system-and-commands-of-linux.html"><a href="file-system-and-commands-of-linux.html#check-you-job-status"><i class="fa fa-check"></i><b>3.9</b> Check you job status</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html"><i class="fa fa-check"></i><b>4</b> Achiving and compressing files</a><ul>
<li class="chapter" data-level="4.1" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#common-compressed-file-format"><i class="fa fa-check"></i><b>4.1</b> Common compressed file format</a></li>
<li class="chapter" data-level="4.2" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#how-to-work-with-different-format"><i class="fa fa-check"></i><b>4.2</b> How to work with different format</a><ul>
<li class="chapter" data-level="4.2.1" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#gz"><i class="fa fa-check"></i><b>4.2.1</b> <code>*.gz</code></a></li>
<li class="chapter" data-level="4.2.2" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#tar.gz"><i class="fa fa-check"></i><b>4.2.2</b> <code>*.tar.gz</code></a></li>
<li class="chapter" data-level="4.2.3" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#zip"><i class="fa fa-check"></i><b>4.2.3</b> <code>*.zip</code></a></li>
<li class="chapter" data-level="4.2.4" data-path="achiving-and-compressing-files.html"><a href="achiving-and-compressing-files.html#tar.bz2"><i class="fa fa-check"></i><b>4.2.4</b> <code>*.tar.bz2</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html"><i class="fa fa-check"></i><b>5</b> Process management in Linux</a><ul>
<li class="chapter" data-level="5.1" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#top"><i class="fa fa-check"></i><b>5.1</b> <code>top</code></a></li>
<li class="chapter" data-level="5.2" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#ps"><i class="fa fa-check"></i><b>5.2</b> <code>ps</code></a></li>
<li class="chapter" data-level="5.3" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#kill"><i class="fa fa-check"></i><b>5.3</b> <code>kill</code></a></li>
<li class="chapter" data-level="5.4" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#df"><i class="fa fa-check"></i><b>5.4</b> <code>df</code></a></li>
<li class="chapter" data-level="5.5" data-path="process-management-in-linux.html"><a href="process-management-in-linux.html#advanced-topic-free"><i class="fa fa-check"></i><b>5.5</b> Advanced topic <code>free</code></a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="file-transfer.html"><a href="file-transfer.html"><i class="fa fa-check"></i><b>6</b> File transfer</a><ul>
<li class="chapter" data-level="6.1" data-path="file-transfer.html"><a href="file-transfer.html#transferring-files-between-local-computer-and-linux-server"><i class="fa fa-check"></i><b>6.1</b> Transferring files between local computer and Linux server</a><ul>
<li class="chapter" data-level="6.1.1" data-path="file-transfer.html"><a href="file-transfer.html#use-command-line-tools"><i class="fa fa-check"></i><b>6.1.1</b> Use command line tools</a></li>
<li class="chapter" data-level="6.1.2" data-path="file-transfer.html"><a href="file-transfer.html#download-files"><i class="fa fa-check"></i><b>6.1.2</b> Download files</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html"><i class="fa fa-check"></i><b>7</b> Install Bioinformatics software in Linux</a><ul>
<li class="chapter" data-level="7.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installation-from-source-code"><i class="fa fa-check"></i><b>7.1</b> Installation from source code</a><ul>
<li class="chapter" data-level="7.1.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa"><i class="fa fa-check"></i><b>7.1.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="7.1.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-samtools"><i class="fa fa-check"></i><b>7.1.2</b> Install <code>samtools</code></a></li>
<li class="chapter" data-level="7.1.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#align-reads-to-genome-using-bwa-and-store-the-alignment-results-in-sambam-files"><i class="fa fa-check"></i><b>7.1.3</b> Align reads to genome using <code>bwa</code> and store the alignment results in SAM/BAM files</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#installing-a-precompiled-binary-executable"><i class="fa fa-check"></i><b>7.2</b> Installing a precompiled binary (executable)</a><ul>
<li class="chapter" data-level="7.2.1" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-bwa-1"><i class="fa fa-check"></i><b>7.2.1</b> Install <code>bwa</code></a></li>
<li class="chapter" data-level="7.2.2" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-with-conda-recommended-way"><i class="fa fa-check"></i><b>7.2.2</b> Install with <code>conda</code> (recommended way)</a></li>
<li class="chapter" data-level="7.2.3" data-path="install-bioinformatics-software-in-linux.html"><a href="install-bioinformatics-software-in-linux.html#install-ussing-docker-advanced-topic"><i class="fa fa-check"></i><b>7.2.3</b> Install ussing <code>Docker</code> (Advanced topic)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html"><i class="fa fa-check"></i><b>8</b> Text editor in Linux</a><ul>
<li class="chapter" data-level="8.1" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#basic-vi-skills"><i class="fa fa-check"></i><b>8.1</b> Basic vi skills</a></li>
<li class="chapter" data-level="8.2" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#create-new-text-file-with-vi"><i class="fa fa-check"></i><b>8.2</b> Create new text file with <code>vi</code></a></li>
<li class="chapter" data-level="8.3" data-path="text-editor-in-linux.html"><a href="text-editor-in-linux.html#an-example-for-using-editor-r"><i class="fa fa-check"></i><b>8.3</b> An example for using editor R</a></li>
</ul></li>
<li class="part"><span><b>II Perl</b></span></li>
<li class="chapter" data-level="9" data-path="first-perl-program.html"><a href="first-perl-program.html"><i class="fa fa-check"></i><b>9</b> First Perl Program</a><ul>
<li class="chapter" data-level="9.1" data-path="first-perl-program.html"><a href="first-perl-program.html#first-program"><i class="fa fa-check"></i><b>9.1</b> First Program</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="varible-in-perl.html"><a href="varible-in-perl.html"><i class="fa fa-check"></i><b>10</b> Varible in Perl</a><ul>
<li class="chapter" data-level="10.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#scalar-variable"><i class="fa fa-check"></i><b>10.1</b> Scalar variable</a></li>
<li class="chapter" data-level="10.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#arithmetic-operations-in-perl"><i class="fa fa-check"></i><b>10.2</b> Arithmetic operations in Perl</a><ul>
<li class="chapter" data-level="10.2.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#shorthand-operations"><i class="fa fa-check"></i><b>10.2.1</b> Shorthand operations</a></li>
<li class="chapter" data-level="10.2.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#auto-increment-and-auto-decrement"><i class="fa fa-check"></i><b>10.2.2</b> Auto increment and auto decrement</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-strict-user-warnings-and-my"><i class="fa fa-check"></i><b>10.3</b> <code>use strict</code>; <code>user warnings</code> and <code>my</code></a></li>
<li class="chapter" data-level="10.4" data-path="varible-in-perl.html"><a href="varible-in-perl.html#array"><i class="fa fa-check"></i><b>10.4</b> Array</a><ul>
<li class="chapter" data-level="10.4.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#init-an-array"><i class="fa fa-check"></i><b>10.4.1</b> Init an array</a></li>
<li class="chapter" data-level="10.4.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#array-index"><i class="fa fa-check"></i><b>10.4.2</b> Array index</a></li>
<li class="chapter" data-level="10.4.3" data-path="varible-in-perl.html"><a href="varible-in-perl.html#length-of-the-array"><i class="fa fa-check"></i><b>10.4.3</b> Length of the array</a></li>
<li class="chapter" data-level="10.4.4" data-path="varible-in-perl.html"><a href="varible-in-perl.html#sort-arrays-in-perl"><i class="fa fa-check"></i><b>10.4.4</b> Sort Arrays in Perl</a></li>
<li class="chapter" data-level="10.4.5" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-push-pop-shift-and-unshift-in-perl"><i class="fa fa-check"></i><b>10.4.5</b> Use <code>push</code>, <code>pop</code>, <code>shift</code> and <code>unshift</code> in Perl</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="varible-in-perl.html"><a href="varible-in-perl.html#hash-in-perl"><i class="fa fa-check"></i><b>10.5</b> Hash in Perl</a><ul>
<li class="chapter" data-level="10.5.1" data-path="varible-in-perl.html"><a href="varible-in-perl.html#sort-keys-numerically-using-sort-ab"><i class="fa fa-check"></i><b>10.5.1</b> Sort keys numerically using <code>sort {$a&lt;=&gt;$b}</code></a></li>
<li class="chapter" data-level="10.5.2" data-path="varible-in-perl.html"><a href="varible-in-perl.html#use-exists-function-on-a-hash"><i class="fa fa-check"></i><b>10.5.2</b> Use <code>exists</code> function on a hash</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="control-structure.html"><a href="control-structure.html"><i class="fa fa-check"></i><b>11</b> Control structure</a><ul>
<li class="chapter" data-level="11.1" data-path="control-structure.html"><a href="control-structure.html#for-loop"><i class="fa fa-check"></i><b>11.1</b> <code>for</code> loop</a></li>
<li class="chapter" data-level="11.2" data-path="control-structure.html"><a href="control-structure.html#foreach-loop"><i class="fa fa-check"></i><b>11.2</b> <code>foreach</code> loop</a></li>
<li class="chapter" data-level="11.3" data-path="control-structure.html"><a href="control-structure.html#while-loop"><i class="fa fa-check"></i><b>11.3</b> <code>while</code> loop</a></li>
<li class="chapter" data-level="11.4" data-path="control-structure.html"><a href="control-structure.html#statement-if-else"><i class="fa fa-check"></i><b>11.4</b> Statement if-else</a></li>
<li class="chapter" data-level="11.5" data-path="control-structure.html"><a href="control-structure.html#operator-last-and-next"><i class="fa fa-check"></i><b>11.5</b> Operator <code>last</code> and <code>next</code></a></li>
<li class="chapter" data-level="11.6" data-path="control-structure.html"><a href="control-structure.html#operator-redo"><i class="fa fa-check"></i><b>11.6</b> Operator <code>redo</code></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="string-manipulation.html"><a href="string-manipulation.html"><i class="fa fa-check"></i><b>12</b> String manipulation</a><ul>
<li class="chapter" data-level="12.1" data-path="string-manipulation.html"><a href="string-manipulation.html#string-concatenation"><i class="fa fa-check"></i><b>12.1</b> String concatenation</a></li>
<li class="chapter" data-level="12.2" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-extraction"><i class="fa fa-check"></i><b>12.2</b> Substring extraction</a></li>
<li class="chapter" data-level="12.3" data-path="string-manipulation.html"><a href="string-manipulation.html#substring-search"><i class="fa fa-check"></i><b>12.3</b> Substring search</a></li>
<li class="chapter" data-level="12.4" data-path="string-manipulation.html"><a href="string-manipulation.html#split-string"><i class="fa fa-check"></i><b>12.4</b> Split String</a></li>
<li class="chapter" data-level="12.5" data-path="string-manipulation.html"><a href="string-manipulation.html#regular-expression"><i class="fa fa-check"></i><b>12.5</b> Regular expression</a><ul>
<li class="chapter" data-level="12.5.1" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-matching"><i class="fa fa-check"></i><b>12.5.1</b> Pattern matching</a></li>
<li class="chapter" data-level="12.5.2" data-path="string-manipulation.html"><a href="string-manipulation.html#pattern-substitution"><i class="fa fa-check"></i><b>12.5.2</b> Pattern substitution</a></li>
<li class="chapter" data-level="12.5.3" data-path="string-manipulation.html"><a href="string-manipulation.html#modifiers-to-pattern-matching-and-substitution"><i class="fa fa-check"></i><b>12.5.3</b> Modifiers to pattern matching and substitution</a></li>
<li class="chapter" data-level="12.5.4" data-path="string-manipulation.html"><a href="string-manipulation.html#greedy-or-non-greedy"><i class="fa fa-check"></i><b>12.5.4</b> Greedy or non-greedy</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html"><i class="fa fa-check"></i><b>13</b> Input and output in Perl</a><ul>
<li class="chapter" data-level="13.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input"><i class="fa fa-check"></i><b>13.1</b> Input</a><ul>
<li class="chapter" data-level="13.1.1" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#standard-input"><i class="fa fa-check"></i><b>13.1.1</b> Standard input</a></li>
<li class="chapter" data-level="13.1.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#input-from-a-file-on-the-disk"><i class="fa fa-check"></i><b>13.1.2</b> Input from a file on the disk</a></li>
<li class="chapter" data-level="13.1.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#special-variable-_"><i class="fa fa-check"></i><b>13.1.3</b> Special variable <code>$_</code></a></li>
<li class="chapter" data-level="13.1.4" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#the-input-record-separator"><i class="fa fa-check"></i><b>13.1.4</b> The input record separator: <code>$/</code></a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#output-to-a-file-on-the-disk"><i class="fa fa-check"></i><b>13.2</b> Output to a file on the disk</a></li>
<li class="chapter" data-level="13.3" data-path="input-and-output-in-perl.html"><a href="input-and-output-in-perl.html#arguments-from-command-line"><i class="fa fa-check"></i><b>13.3</b> Arguments from command line</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="practical-perl-program.html"><a href="practical-perl-program.html"><i class="fa fa-check"></i><b>14</b> Practical Perl program</a><ul>
<li class="chapter" data-level="14.1" data-path="practical-perl-program.html"><a href="practical-perl-program.html#add-annotation-information-to-deseq2-results"><i class="fa fa-check"></i><b>14.1</b> Add annotation information to DESeq2 results</a></li>
<li class="chapter" data-level="14.2" data-path="practical-perl-program.html"><a href="practical-perl-program.html#merge-overlap-genomic-regions"><i class="fa fa-check"></i><b>14.2</b> Merge overlap genomic regions</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="perl-modules.html"><a href="perl-modules.html"><i class="fa fa-check"></i><b>15</b> Perl modules</a><ul>
<li class="chapter" data-level="15.1" data-path="perl-modules.html"><a href="perl-modules.html#what-is-a-perl-module"><i class="fa fa-check"></i><b>15.1</b> What is a Perl module</a></li>
<li class="chapter" data-level="15.2" data-path="perl-modules.html"><a href="perl-modules.html#how-to-install-a-perl-module"><i class="fa fa-check"></i><b>15.2</b> How to install a Perl module</a></li>
<li class="chapter" data-level="15.3" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-a-perl-module"><i class="fa fa-check"></i><b>15.3</b> How to use a Perl module</a></li>
<li class="chapter" data-level="15.4" data-path="perl-modules.html"><a href="perl-modules.html#how-to-use-bioperl-module"><i class="fa fa-check"></i><b>15.4</b> How to use BioPerl module</a></li>
<li class="chapter" data-level="15.5" data-path="perl-modules.html"><a href="perl-modules.html#how-to-write-a-module"><i class="fa fa-check"></i><b>15.5</b> How to write a module</a></li>
</ul></li>
<li class="part"><span><b>III Statistics and R</b></span></li>
<li class="chapter" data-level="16" data-path="r-introduction.html"><a href="r-introduction.html"><i class="fa fa-check"></i><b>16</b> R introduction</a><ul>
<li class="chapter" data-level="16.1" data-path="r-introduction.html"><a href="r-introduction.html#basic-r-function"><i class="fa fa-check"></i><b>16.1</b> Basic R function</a></li>
<li class="chapter" data-level="16.2" data-path="r-introduction.html"><a href="r-introduction.html#producing-simple-graphs-with-r"><i class="fa fa-check"></i><b>16.2</b> Producing Simple Graphs with R</a><ul>
<li class="chapter" data-level="16.2.1" data-path="r-introduction.html"><a href="r-introduction.html#line-charts"><i class="fa fa-check"></i><b>16.2.1</b> Line Charts</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html"><i class="fa fa-check"></i><b>17</b> Producing simple graphs</a><ul>
<li class="chapter" data-level="17.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#producing-simple-graphs-using-r"><i class="fa fa-check"></i><b>17.1</b> Producing simple graphs using R</a><ul>
<li class="chapter" data-level="17.1.1" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#line-charts-1"><i class="fa fa-check"></i><b>17.1.1</b> Line Charts</a></li>
<li class="chapter" data-level="17.1.2" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#bar-charts"><i class="fa fa-check"></i><b>17.1.2</b> Bar Charts</a></li>
<li class="chapter" data-level="17.1.3" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#histograms"><i class="fa fa-check"></i><b>17.1.3</b> Histograms</a></li>
<li class="chapter" data-level="17.1.4" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#pie-charts"><i class="fa fa-check"></i><b>17.1.4</b> Pie Charts</a></li>
<li class="chapter" data-level="17.1.5" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#dotcharts"><i class="fa fa-check"></i><b>17.1.5</b> Dotcharts</a></li>
<li class="chapter" data-level="17.1.6" data-path="producing-simple-graphs.html"><a href="producing-simple-graphs.html#misc"><i class="fa fa-check"></i><b>17.1.6</b> Misc</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="18" data-path="ggplot2.html"><a href="ggplot2.html"><i class="fa fa-check"></i><b>18</b> <code>ggplot2</code></a><ul>
<li class="chapter" data-level="18.1" data-path="ggplot2.html"><a href="ggplot2.html#ggplot2-1"><i class="fa fa-check"></i><b>18.1</b> ggplot2</a></li>
<li class="chapter" data-level="18.2" data-path="ggplot2.html"><a href="ggplot2.html#ggplot2-practical"><i class="fa fa-check"></i><b>18.2</b> ggplot2 practical</a></li>
</ul></li>
<li class="chapter" data-level="19" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html"><i class="fa fa-check"></i><b>19</b> Heatmap Tutorial</a><ul>
<li class="chapter" data-level="19.1" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#install-pheatmap-package"><i class="fa fa-check"></i><b>19.1</b> Install <code>pheatmap</code> package</a></li>
<li class="chapter" data-level="19.2" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#draw-a-heatmap-for-gene-expression-of-rna-seq-data"><i class="fa fa-check"></i><b>19.2</b> Draw a heatmap for gene expression of RNA-seq data</a></li>
<li class="chapter" data-level="19.3" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#add-the-annotation"><i class="fa fa-check"></i><b>19.3</b> Add the annotation</a></li>
<li class="chapter" data-level="19.4" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section"><i class="fa fa-check"></i><b>19.4</b> </a></li>
<li class="chapter" data-level="19.5" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#transfrom-the-data"><i class="fa fa-check"></i><b>19.5</b> Transfrom the data</a></li>
<li class="chapter" data-level="19.6" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-add-annotations"><i class="fa fa-check"></i><b>19.6</b> How to add annotations</a></li>
<li class="chapter" data-level="19.7" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-cut-the-trees"><i class="fa fa-check"></i><b>19.7</b> How to cut the trees</a></li>
<li class="chapter" data-level="19.8" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#how-to-get-the-cluster-information-from-the-heatmap"><i class="fa fa-check"></i><b>19.8</b> How to get the cluster information from the heatmap</a></li>
<li class="chapter" data-level="19.9" data-path="heatmap-tutorial.html"><a href="heatmap-tutorial.html#section-1"><i class="fa fa-check"></i><b>19.9</b> </a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html"><i class="fa fa-check"></i><b>20</b> Preparation of figures for manuscript</a><ul>
<li class="chapter" data-level="20.1" data-path="preparation-of-figures-for-manuscript.html"><a href="preparation-of-figures-for-manuscript.html#test1"><i class="fa fa-check"></i><b>20.1</b> test1</a></li>
</ul></li>
<li class="part"><span><b>IV Omics data</b></span></li>
<li class="chapter" data-level="21" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html"><i class="fa fa-check"></i><b>21</b> Introduction to NGS</a><ul>
<li class="chapter" data-level="21.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#introduction-of-biology-for-bioinformatics"><i class="fa fa-check"></i><b>21.1</b> Introduction of Biology for Bioinformatics</a><ul>
<li class="chapter" data-level="21.1.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-dna"><i class="fa fa-check"></i><b>21.1.1</b> What is DNA</a></li>
<li class="chapter" data-level="21.1.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-genome"><i class="fa fa-check"></i><b>21.1.2</b> What is genome</a></li>
<li class="chapter" data-level="21.1.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#how-to-assemble-a-genome"><i class="fa fa-check"></i><b>21.1.3</b> How to assemble a genome</a></li>
<li class="chapter" data-level="21.1.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-gene"><i class="fa fa-check"></i><b>21.1.4</b> What is gene</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#what-is-ngs"><i class="fa fa-check"></i><b>21.2</b> What is NGS</a></li>
<li class="chapter" data-level="21.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#application-of-ngs"><i class="fa fa-check"></i><b>21.3</b> Application of NGS</a></li>
<li class="chapter" data-level="21.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#section-2"><i class="fa fa-check"></i><b>21.4</b> </a></li>
<li class="chapter" data-level="21.5" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#data-file-formats-to-store-the-genomic-information"><i class="fa fa-check"></i><b>21.5</b> Data file formats to store the genomic information</a><ul>
<li class="chapter" data-level="21.5.1" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#fasta-file"><i class="fa fa-check"></i><b>21.5.1</b> <code>fasta</code> file</a></li>
<li class="chapter" data-level="21.5.2" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#gff-file"><i class="fa fa-check"></i><b>21.5.2</b> GFF file</a></li>
<li class="chapter" data-level="21.5.3" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#gtf-file"><i class="fa fa-check"></i><b>21.5.3</b> GTF file</a></li>
<li class="chapter" data-level="21.5.4" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#fastq-file"><i class="fa fa-check"></i><b>21.5.4</b> FASTQ file</a></li>
<li class="chapter" data-level="21.5.5" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#sambam-file"><i class="fa fa-check"></i><b>21.5.5</b> SAM/BAM file</a></li>
<li class="chapter" data-level="21.5.6" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#vcf-file"><i class="fa fa-check"></i><b>21.5.6</b> VCF file</a></li>
<li class="chapter" data-level="21.5.7" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#bed-format"><i class="fa fa-check"></i><b>21.5.7</b> BED format</a></li>
<li class="chapter" data-level="21.5.8" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#bedgraph-format"><i class="fa fa-check"></i><b>21.5.8</b> bedGraph format</a></li>
</ul></li>
<li class="chapter" data-level="21.6" data-path="introduction-to-ngs.html"><a href="introduction-to-ngs.html#usefull-links"><i class="fa fa-check"></i><b>21.6</b> Usefull links:</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="bs-seq.html"><a href="bs-seq.html"><i class="fa fa-check"></i><b>22</b> BS-seq</a><ul>
<li class="chapter" data-level="22.1" data-path="bs-seq.html"><a href="bs-seq.html#section-3"><i class="fa fa-check"></i><b>22.1</b> </a></li>
<li class="chapter" data-level="22.2" data-path="bs-seq.html"><a href="bs-seq.html#what-is-bs-seq"><i class="fa fa-check"></i><b>22.2</b> What is BS-seq</a></li>
<li class="chapter" data-level="22.3" data-path="bs-seq.html"><a href="bs-seq.html#section-4"><i class="fa fa-check"></i><b>22.3</b> </a></li>
<li class="chapter" data-level="22.4" data-path="bs-seq.html"><a href="bs-seq.html#how-to-analyze-bs-seq-data"><i class="fa fa-check"></i><b>22.4</b> How to analyze BS-seq data</a><ul>
<li class="chapter" data-level="22.4.1" data-path="bs-seq.html"><a href="bs-seq.html#section-5"><i class="fa fa-check"></i><b>22.4.1</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="start-a-project.html"><a href="start-a-project.html"><i class="fa fa-check"></i><b>23</b> Start a project</a><ul>
<li class="chapter" data-level="23.1" data-path="start-a-project.html"><a href="start-a-project.html#experimetal-design"><i class="fa fa-check"></i><b>23.1</b> Experimetal design</a><ul>
<li class="chapter" data-level="23.1.1" data-path="start-a-project.html"><a href="start-a-project.html#how-many-biological-replicates"><i class="fa fa-check"></i><b>23.1.1</b> How many biological replicates</a></li>
<li class="chapter" data-level="23.1.2" data-path="start-a-project.html"><a href="start-a-project.html#how-big"><i class="fa fa-check"></i><b>23.1.2</b> How big</a></li>
<li class="chapter" data-level="23.1.3" data-path="start-a-project.html"><a href="start-a-project.html#how-much-data-to-generate"><i class="fa fa-check"></i><b>23.1.3</b> How much data to generate</a></li>
<li class="chapter" data-level="23.1.4" data-path="start-a-project.html"><a href="start-a-project.html#section-6"><i class="fa fa-check"></i><b>23.1.4</b> </a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="24" data-path="capstone-project.html"><a href="capstone-project.html"><i class="fa fa-check"></i><b>24</b> Capstone project:</a><ul>
<li class="chapter" data-level="24.1" data-path="capstone-project.html"><a href="capstone-project.html#introduction"><i class="fa fa-check"></i><b>24.1</b> Introduction</a></li>
<li class="chapter" data-level="24.2" data-path="capstone-project.html"><a href="capstone-project.html#method"><i class="fa fa-check"></i><b>24.2</b> Method</a></li>
<li class="chapter" data-level="24.3" data-path="capstone-project.html"><a href="capstone-project.html#pipelines"><i class="fa fa-check"></i><b>24.3</b> Pipelines</a></li>
<li class="chapter" data-level="24.4" data-path="capstone-project.html"><a href="capstone-project.html#section-7"><i class="fa fa-check"></i><b>24.4</b> </a></li>
</ul></li>
<li class="chapter" data-level="25" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html"><i class="fa fa-check"></i><b>25</b> Practical data analysis on WGBS</a><ul>
<li class="chapter" data-level="25.1" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#donwload-the-data"><i class="fa fa-check"></i><b>25.1</b> Donwload the data</a><ul>
<li class="chapter" data-level="25.1.1" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#reference"><i class="fa fa-check"></i><b>25.1.1</b> Reference</a></li>
<li class="chapter" data-level="25.1.2" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#now-lets-download-the-wgbs-data"><i class="fa fa-check"></i><b>25.1.2</b> Now let’s download the WGBS data</a></li>
</ul></li>
<li class="chapter" data-level="25.2" data-path="practical-data-analysis-on-wgbs.html"><a href="practical-data-analysis-on-wgbs.html#convert-sra-files-to-fastq-files"><i class="fa fa-check"></i><b>25.2</b> Convert <code>sra</code> files to <code>fastq</code> files</a></li>
</ul></li>
<li class="chapter" data-level="26" data-path="section-8.html"><a href="section-8.html"><i class="fa fa-check"></i><b>26</b> </a><ul>
<li class="chapter" data-level="26.1" data-path="section-8.html"><a href="section-8.html#including-plots"><i class="fa fa-check"></i><b>26.1</b> Including Plots</a></li>
</ul></li>
<li class="chapter" data-level="27" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html"><i class="fa fa-check"></i><b>27</b> Good resouces to learn Bioinformatics</a><ul>
<li class="chapter" data-level="27.1" data-path="good-resouces-to-learn-bioinformatics.html"><a href="good-resouces-to-learn-bioinformatics.html#references"><i class="fa fa-check"></i><b>27.1</b> References:</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references-1.html"><a href="references-1.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="28" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html"><i class="fa fa-check"></i><b>28</b> CRISPR-CAS9 screen data processing</a><ul>
<li class="chapter" data-level="28.1" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#installation-of-softwares-using-conda"><i class="fa fa-check"></i><b>28.1</b> Installation of softwares using <code>conda</code></a></li>
<li class="chapter" data-level="28.2" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#an-exmaple-of-how-to-process"><i class="fa fa-check"></i><b>28.2</b> An exmaple of how to process</a><ul>
<li class="chapter" data-level="28.2.1" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#donwload-the-data-1"><i class="fa fa-check"></i><b>28.2.1</b> Donwload the data</a></li>
<li class="chapter" data-level="28.2.2" data-path="crispr-cas9-screen-data-processing.html"><a href="crispr-cas9-screen-data-processing.html#reference-1"><i class="fa fa-check"></i><b>28.2.2</b> Reference</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="29" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html"><i class="fa fa-check"></i><b>29</b> Commands for Linux administration (Advanced topic)</a><ul>
<li class="chapter" data-level="29.1" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#w"><i class="fa fa-check"></i><b>29.1</b> <code>w</code></a></li>
<li class="chapter" data-level="29.2" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#who"><i class="fa fa-check"></i><b>29.2</b> <code>who</code></a></li>
<li class="chapter" data-level="29.3" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#uptime"><i class="fa fa-check"></i><b>29.3</b> <code>uptime</code></a></li>
<li class="chapter" data-level="29.4" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#whoami"><i class="fa fa-check"></i><b>29.4</b> <code>whoami</code></a></li>
<li class="chapter" data-level="29.5" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#ifconfig"><i class="fa fa-check"></i><b>29.5</b> <code>ifconfig</code></a></li>
<li class="chapter" data-level="29.6" data-path="commands-for-linux-administration-advanced-topic.html"><a href="commands-for-linux-administration-advanced-topic.html#useradd-and-passwd"><i class="fa fa-check"></i><b>29.6</b> <code>useradd</code> and <code>passwd</code></a></li>
</ul></li>
<li class="part"><span><b>V Python</b></span></li>
<li class="chapter" data-level="30" data-path="first-perl-program-1.html"><a href="first-perl-program-1.html"><i class="fa fa-check"></i><b>30</b> First Perl Program</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">All things considered for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="varible-in-perl" class="section level1">
<h1><span class="header-section-number">Chapter 10</span> Varible in Perl</h1>
<p>Perl provides three kinds of variables: <code>scalars</code>, <code>arrays</code>, and <code>hash</code>(aka <code>associative arrays</code>). The initial character of the name identifies the particular type of variable and, hence, its functionality.</p>
<table>
<thead>
<tr class="header">
<th>Type</th>
<th>Character</th>
<th>Example</th>
<th>Is a name for:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Scalar</td>
<td>$</td>
<td>$length</td>
<td>An individual value (number or string)</td>
</tr>
<tr class="even">
<td>Array</td>
<td>@</td>
<td>@gene_list</td>
<td>A list of values, keyed by number</td>
</tr>
<tr class="odd">
<td>Hash</td>
<td>%</td>
<td>%gene_annotation</td>
<td>A group of values, keyed by string</td>
</tr>
</tbody>
</table>
<div id="scalar-variable" class="section level2">
<h2><span class="header-section-number">10.1</span> Scalar variable</h2>
<p>In Perl, scalar variables can be used to store mainly two types of data: string and numbers.</p>
<p>Perl does not differentiate between a number and a string, nor does it differentiate between integers and reals.</p>
<p>In order to tell the computer what to print, we need to use variables. In Perl, the name of a <code>scalar</code> variable starts with the dollar sign <code>$</code>. You can assign either a number or a string to it.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="co">#assign two strings to two variables</span>
<span class="kw">my</span> <span class="dt">$dna_seq1</span> = <span class="kw">&quot;</span><span class="st">ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna_seq2</span> = <span class="kw">&quot;</span><span class="st">TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna_seq1</span>;
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna_seq2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of first DNA sequence (</span><span class="dt">$dna_seq1</span><span class="st">) is: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of the Second DNA sequence (</span><span class="dt">$dna_seq2</span><span class="st">) is: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/variable_assign.pl</code></pre>
<pre><code>## The length of first DNA sequence (ACCTCGGTACAGTGAATGGGAAACGTAGCTGAT) is: 33
## The length of the Second DNA sequence (TGCCGATCGTAATAGCTCGCTATCTAGCTCGATCGTCGTA) is: 40</code></pre>
<p>Here are the explanations of this script.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="co"># assign two DNA sequences to two variables</span>
<span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;</code></pre>
<p>The first line here is a comment. In Perl, <code>#</code> (pound sign) is the comment character. The comments can be used to document the program and improve the readability. During the execution, the comments will be ignored. The second and third lines declare two string scalar variables (<code>$dna1</code> and <code>$dna2</code>) to store two DNA sequences.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna_seq1</span>; 
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna_seq2</span>;</code></pre>
<p>The code above has one comment line and declares two integer scalar variables to store the return values of built-on function <code>length</code>.</p>
<p>Now you can output the variables to see what are stored in them using the following code:</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of first DNA sequence (</span><span class="dt">$dna_seq1</span><span class="st">) is: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length of the Second DNA sequence (</span><span class="dt">$dna_seq2</span><span class="st">) is: </span><span class="dt">$dna_length2</span><span class="ch">\n\n</span><span class="kw">&quot;</span>;</code></pre>
<p>This script shows you what string scalar varibles and integer scalr varibles.</p>
</div>
<div id="arithmetic-operations-in-perl" class="section level2">
<h2><span class="header-section-number">10.2</span> Arithmetic operations in Perl</h2>
<p>Most arithmetic operators are binary operators; this means they take two arguments. Unary operators only take one argument. Arithmetic operators are very simple and often transparent.</p>
<p>Here we’re mainly going to talk about basic arithmetic opertators including addition (<code>+</code>), substraction (<code>-</code>), multiplication (<code>*</code>), division (<code>/</code>) and the modulus operation (%). Modulus (<code>%</code>) returns the remainder of a division (<code>/</code>) operation.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="co">## If we know the lengths of two sequences and we want to calulate the sum of the lengths.</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="dv">100</span>;
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="dv">200</span>;
<span class="kw">my</span> <span class="dt">$total_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_length2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of these two sequences is </span><span class="dt">$total_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="kw">my</span> <span class="dt">$length_diff</span> = <span class="dt">$dna_length1</span> - <span class="dt">$dna_length2</span>; 
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The length difference is </span><span class="dt">$length_diff</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="kw">my</span> <span class="dt">$average_length</span> = <span class="dt">$total_length</span> / <span class="dv">2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The average length of two DNA sequences is </span><span class="dt">$average_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## The total length of these two sequences is 300 
## The length difference is -100
## The average length of two DNA sequences is 150</code></pre>
<p>The code above shows us how to</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="co">#Imaging the CC content of first DNA sequence is 0.5.</span>
<span class="co">#How many CC do we have in first DNA?</span>
<span class="kw">my</span> <span class="dt">$cg_content</span> = <span class="fl">0.5</span>;
<span class="kw">my</span> <span class="dt">$cg_number</span> = <span class="dt">$dna_length1</span> <span class="kw">*</span> <span class="fl">0.5</span>;</code></pre>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="co">#Imaging we have a 10 bps DNA sequnces,</span>
<span class="co">#how many possible DNA sequences do we have?</span>
<span class="kw">my</span> <span class="dt">$dna_nucleotide</span> = <span class="kw">&quot;</span><span class="st">ATCG</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna_nucleo_number</span> = <span class="fu">length</span> <span class="dt">$dna_nucleotide</span>;
<span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;
<span class="kw">my</span> <span class="dt">$possible_number</span> = <span class="dt">$dna_nucleo_number</span> *<span class="kw">*</span> <span class="dt">$dna_length</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">We have </span><span class="dt">$possible_number</span><span class="st"> possibilities.</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## We have 1048576 possibilities.</code></pre>
<div id="shorthand-operations" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Shorthand operations</h3>
<p>The expression $x += 3; is the shorthand version of $x = $x + 3;, they have exactly the same result:</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="fu">use</span> <span class="kw">strict</span>;
<span class="fu">use</span> <span class="kw">warnings</span>;

<span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="dt">$dna_length</span> += <span class="dv">3</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length += 3&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="dt">$dna_length</span> -= <span class="dv">3</span>;  
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length -= 3&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## DNA length: 10
## DNA length: 13 after &#39;$dna_length += 3&#39;
## DNA length: 10 after &#39;$dna_length -= 3&#39;</code></pre>
</div>
<div id="auto-increment-and-auto-decrement" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Auto increment and auto decrement</h3>
<p><code>++</code> and <code>--</code> are provided for the auto increment and auto decrement operators. They increase and decrease respectively the value of a scalar variable by 1.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="fu">use</span> <span class="kw">strict</span>;
<span class="fu">use</span> <span class="kw">warnings</span>;

<span class="kw">my</span> <span class="dt">$dna_length</span> = <span class="dv">10</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="dt">$dna_length</span> ++;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after &#39;\$dna_length ++&#39;</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="dt">$dna_length</span> --;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">DNA length: </span><span class="dt">$dna_length</span><span class="st"> after `\$dna_length --`</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## DNA length: 10
## DNA length: 11 after &#39;$dna_length ++&#39;
## DNA length: 10 after `$dna_length --`</code></pre>
</div>
</div>
<div id="use-strict-user-warnings-and-my" class="section level2">
<h2><span class="header-section-number">10.3</span> <code>use strict</code>; <code>user warnings</code> and <code>my</code></h2>
<p>For starters, <code>use strict;</code> (and to a lesser extent, <code>use warnings;</code>) helps find typos in variable names. Even experienced programmers make such errors. A common case is forgetting to rename an instance of a variable when cleaning up or refactoring code.</p>
<p>Using <code>use strict; use warnings;</code> catches many errors sooner than they would be caught otherwise, which makes it easier to find the root causes of the errors. The root cause might be the need for an error or validation check, and that can happen regardless of programmer skill.</p>
<p>What’s good about Perl warnings is that they are rarely spurious, so there’s next to no cost to using them.</p>
<p>In the script below, <code>$dna_lenght2</code> is a typo. If you run this script, it will give you the output without any error message, although it’s not the right output.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>

<span class="co">#assign 2 numbers to 2 variable</span>
<span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;
<span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calculate the total length of two DNA sequences</span>
<span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<p>Let’s try to run this script:</p>
<pre class="sourceCode sh"><code class="sourceCode bash"><span class="fu">perl</span> code_perl/var_assign_no_strict_warnings.pl</code></pre>
<pre><code>First DNA: CTCGACCAGGACGATGAATGGGCGATGAAAATCT; Length: 34
Second DNA: CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT; Length: 48
The total length of two DNA sequences is 34 </code></pre>
<p>So the script above is supposed to output the length of two DNA sequences and the sum of the lengths.</p>
<p>In the chunk of code above, <code>$dna_lenght2</code> is an empty varible without storing any information. By defaut, Perl considers this as ZERO when doing plus operation. Although there was no error message given here, we infact have an incorrect output.</p>
<p>If we add <code>use strict</code> and <code>use warnings</code>, we need to decare each variable in the script. Let us see what will happen if we have an typo.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">#!/usr/bin/perl</span>
<span class="fu">use</span> <span class="kw">warnings</span>;
<span class="fu">use</span> <span class="kw">strict</span>;

<span class="co">#assign 2 numbers to 2 variable</span>
<span class="kw">my</span> <span class="dt">$dna1</span> = <span class="kw">&quot;</span><span class="st">CTCGACCAGGACGATGAATGGGCGATGAAAATCT</span><span class="kw">&quot;</span>;
<span class="kw">my</span> <span class="dt">$dna2</span> = <span class="kw">&quot;</span><span class="st">CGCTAAACGCTAAACCCTAAACGCTAAACCTCTGAATCCTTAATCGCT</span><span class="kw">&quot;</span>;

<span class="co">#Returns the length in characters of the value of EXPR</span>
<span class="kw">my</span> <span class="dt">$dna_length1</span> = <span class="fu">length</span> <span class="dt">$dna1</span>;
<span class="kw">my</span> <span class="dt">$dna_length2</span> = <span class="fu">length</span> <span class="dt">$dna2</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First DNA: </span><span class="dt">$dna1</span><span class="st">; Length: </span><span class="dt">$dna_length1</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Second DNA: </span><span class="dt">$dna2</span><span class="st">; Length: </span><span class="dt">$dna_length2</span><span class="ch">\n</span><span class="kw">&quot;</span>;

<span class="co">#calculate the total length of two DNA sequences</span>
<span class="kw">my</span> <span class="dt">$tot_length</span> = <span class="dt">$dna_length1</span> + <span class="dt">$dna_lenght2</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The total length of two DNA sequences is </span><span class="dt">$tot_length</span><span class="st"> </span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name (did you forget to declare &quot;my $dna_lenght2&quot;?) at code_perl/var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted due to compilation errors.</code></pre>
<p>Now if we run this script, we encounter error mesage and the script can’t be sucessfuly excuted.</p>
<pre><code>Global symbol &quot;$dna_lenght2&quot; requires explicit package name 
(did you forget to declare &quot;my $dna_lenght2&quot;?) at code_perl/
var_assign_strict_warnings.pl line 17.
Execution of code_perl/var_assign_strict_warnings.pl aborted
 due to compilation errors.</code></pre>
</div>
<div id="array" class="section level2">
<h2><span class="header-section-number">10.4</span> Array</h2>
<div id="init-an-array" class="section level3">
<h3><span class="header-section-number">10.4.1</span> Init an array</h3>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);
<span class="fu">print</span> <span class="kw">&quot;&quot;</span>;
<span class="fu">print</span> <span class="dt">@base_pair</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="fu">join</span>(<span class="kw">&quot;</span><span class="ch">\t</span><span class="kw">&quot;</span>, <span class="dt">@base_pair</span>), <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## ATCCG
## A    T   C   C   G</code></pre>
</div>
<div id="array-index" class="section level3">
<h3><span class="header-section-number">10.4.2</span> Array index</h3>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);
<span class="co">### Extract the first element in the array</span>
<span class="kw">my</span> <span class="dt">$first_base</span> = <span class="dt">$base_pair</span>[<span class="dv">0</span>];
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">First base is: </span><span class="dt">$first_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; 

<span class="co">### Extract the last element in the array</span>
<span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[<span class="dv">4</span>];
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base is: </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; 

<span class="co">### Extract the last element in the array using index `-1`</span>
<span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[-<span class="dv">1</span>];
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base using &#39;-1&#39; is: </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; 

<span class="co">### Extract the last element in the array using index `$#`</span>
<span class="kw">my</span> <span class="dt">$last_base</span> = <span class="dt">$base_pair</span>[<span class="dt">$#base_pair</span>];
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Last base using &#39;\$#&#39; is : </span><span class="dt">$last_base</span><span class="ch">\n\n</span><span class="kw">&quot;</span>; </code></pre>
<pre><code>## First base is: A
## 
## Last base is: G
## 
## Last base using &#39;-1&#39; is: G
## 
## Last base using &#39;$#&#39; is : G</code></pre>
</div>
<div id="length-of-the-array" class="section level3">
<h3><span class="header-section-number">10.4.3</span> Length of the array</h3>
<p>To get the length of the array, you can use <code>scalar @array</code> or <code>just array</code>.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@gene_expr</span> = (<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">10</span>);
<span class="kw">my</span> <span class="dt">$len</span> = <span class="fu">scalar</span> <span class="dt">@gene_expr</span>;  <span class="co">## or @gene_expr</span>
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array has: </span><span class="dt">@gene_expr</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Length of array: </span><span class="kw">&quot;</span>, <span class="dt">$len</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## Array has: 1 3 10
## Length of array: 3</code></pre>
<p>Another way is to use <code>$#array</code>. <code>$#array</code> will return the index of the last element. Since the index starts with 0, to get the length we use <code>$#array + 1</code>.</p>
<p>If the length of array is empty, <code>$#array</code> will return <code>-1</code>.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@gene_expr</span> = ();
<span class="kw">my</span> <span class="dt">$len</span> = <span class="dt">$#gene_expr</span> + <span class="dv">1</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array has </span><span class="dt">$len</span><span class="st"> element</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Length of array: </span><span class="kw">&quot;</span>, <span class="dt">$len</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## Array has 0 element
## Length of array: 0</code></pre>
</div>
<div id="sort-arrays-in-perl" class="section level3">
<h3><span class="header-section-number">10.4.4</span> Sort Arrays in Perl</h3>
<div id="sort-alphebetically" class="section level4">
<h4><span class="header-section-number">10.4.4.1</span> Sort alphebetically</h4>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@base_pair</span> = (<span class="kw">&#39;</span><span class="st">A</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">T</span><span class="kw">&#39;</span>, <span class="kw">&#39;</span><span class="st">C</span><span class="kw">&#39;</span>, <span class="kw">&quot;</span><span class="st">C</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="st">G</span><span class="kw">&quot;</span>);
<span class="kw">my</span> <span class="dt">@sorted_bp</span> = <span class="fu">sort</span> <span class="dt">@base_pair</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@base_pair</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_bp</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## Array before sorted: A T C C G
## Array after sorted: A C C G T</code></pre>
</div>
<div id="sort-numerically" class="section level4">
<h4><span class="header-section-number">10.4.4.2</span> Sort numerically</h4>
<p>To sort an array numerically, we use <code>spaceship operator: &lt;=&gt;</code>.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@genome_coor</span> = (<span class="dv">100</span>, <span class="dv">300</span>, <span class="dv">200</span>, <span class="dv">500</span>);
<span class="kw">my</span> <span class="dt">@sorted_coor</span> = <span class="fu">sort</span> {<span class="dt">$a</span> &lt;=&gt; <span class="dt">$b</span>} <span class="dt">@genome_coor</span>;

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@genome_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## Array before sorted: 100 300 200 500
## Array after sorted: 100 200 300 500</code></pre>
<p>Similarly, array can be also sorted numerically in decreasing order.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">@genome_coor</span> = (<span class="dv">100</span>, <span class="dv">300</span>, <span class="dv">200</span>, <span class="dv">500</span>);
<span class="co">## {$a &lt;=&gt; $b} is modified as {$b &lt;=&gt; $a} </span>
<span class="kw">my</span> <span class="dt">@sorted_coor</span> = <span class="fu">sort</span> {<span class="dt">$b</span> &lt;=&gt; <span class="dt">$a</span>} <span class="dt">@genome_coor</span>;  

<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array before sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@genome_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Array after sorted: </span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="dt">@sorted_coor</span><span class="kw">&quot;</span>, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## Array before sorted: 100 300 200 500
## Array after sorted: 500 300 200 100</code></pre>
</div>
</div>
<div id="use-push-pop-shift-and-unshift-in-perl" class="section level3">
<h3><span class="header-section-number">10.4.5</span> Use <code>push</code>, <code>pop</code>, <code>shift</code> and <code>unshift</code> in Perl</h3>
</div>
</div>
<div id="hash-in-perl" class="section level2">
<h2><span class="header-section-number">10.5</span> Hash in Perl</h2>
<p>A Perl hash is defined by key-value pairs. Perl stores elements of a hash in such an optimal way that you can look up its values based on keys very fast.</p>
<p>With the array, you use indices to access its elements. However, you must use descriptive keys to access hash’s element. A hash is sometimes referred to as an associative array.</p>
<p>Like a scalar or an array variable, a hash variable has its own prefix. A hash variable must begin with a percent sign ( %). The prefix % looks like key/value pair so remember this trick to name the hash variables.</p>
<p>The following example defines a simple hash.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>
                );

<span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene1}</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene2}</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="fu">print</span> <span class="kw">&quot;</span><span class="dt">$gene_info</span><span class="st">{gene3}</span><span class="ch">\n</span><span class="kw">&quot;</span>;</code></pre>
<pre><code>## ROS
## WRKY
## WRKY</code></pre>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>
                );
<span class="kw">my</span> <span class="dt">@hash_key</span>  = <span class="fu">keys</span> <span class="dt">%gene_info</span>;
<span class="dt">@hash_key</span> = <span class="fu">sort</span> <span class="dt">@hash_key</span>;
<span class="co">#print &quot;Keys of \%gene_info: &quot;, @hash_key, &quot;\n&quot;;</span>
<span class="fu">print</span> <span class="kw">&quot;</span><span class="st">Keys of \%gene_info: </span><span class="kw">&quot;</span>, <span class="fu">join</span>(<span class="kw">&quot;</span><span class="ch">\t</span><span class="kw">&quot;</span>, <span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%gene_info</span>), <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="co">#print &quot;Keys of \%gene_info: &quot;, keys %gene_info, &quot;\n&quot;;</span></code></pre>
<pre><code>## Keys of %gene_info: gene1    gene2   gene3</code></pre>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>
                );
<span class="kw">foreach</span>(<span class="fu">sort</span> <span class="fu">keys</span> <span class="dt">%gene_info</span>){
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$gene_info</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}</code></pre>
<pre><code>## The value of gene1 is: ROS
## The value of gene2 is: WRKY
## The value of gene3 is: WRKY</code></pre>
<div id="sort-keys-numerically-using-sort-ab" class="section level3">
<h3><span class="header-section-number">10.5.1</span> Sort keys numerically using <code>sort {$a&lt;=&gt;$b}</code></h3>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">%read_dep</span> = (<span class="kw">&quot;</span><span class="st">1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">100</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">200</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">50</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">10</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">20</span><span class="kw">&quot;</span>,
                );

<span class="co">### sort in increasing order</span>
<span class="kw">foreach</span>(<span class="fu">sort</span> {<span class="dt">$a</span>&lt;=&gt;<span class="dt">$b</span>} <span class="fu">keys</span> <span class="dt">%read_dep</span>){
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$read_dep</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}

<span class="fu">print</span> <span class="kw">&quot;</span><span class="ch">\n</span><span class="st">Sort in descending order: </span><span class="ch">\n</span><span class="kw">&quot;</span>;
<span class="co">### sort {$b&lt;=&gt;$a} if you want to sort in descending order python</span>
<span class="kw">foreach</span>(<span class="fu">sort</span> {<span class="dt">$b</span>&lt;=&gt;<span class="dt">$a</span>} <span class="fu">keys</span> <span class="dt">%read_dep</span>){
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The value of </span><span class="dt">$_</span><span class="st"> is: </span><span class="kw">&quot;</span>, <span class="dt">$read_dep</span>{<span class="dt">$_</span>}, <span class="kw">&quot;</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}</code></pre>
<pre><code>## The value of 1 is: 100
## The value of 2 is: 200
## The value of 3 is: 50
## The value of 10 is: 20
## 
## Sort in descending order: 
## The value of 10 is: 20
## The value of 3 is: 50
## The value of 2 is: 200
## The value of 1 is: 100</code></pre>
</div>
<div id="use-exists-function-on-a-hash" class="section level3">
<h3><span class="header-section-number">10.5.2</span> Use <code>exists</code> function on a hash</h3>
<p>Given an expression that specifies an element of a hash, <code>exists</code> returns true if the specified element in the hash has ever been initialized, even if the corresponding value is undefined.</p>
<pre class="sourceCode perl"><code class="sourceCode perl"><span class="kw">my</span> <span class="dt">%gene_info</span> = (<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">ROS</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene2</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>,
                 <span class="kw">&quot;</span><span class="st">gene3</span><span class="kw">&quot;</span>=&gt;<span class="kw">&quot;</span><span class="st">WRKY</span><span class="kw">&quot;</span>
                );
<span class="kw">if</span>(<span class="fu">exists</span> <span class="dt">$gene_info</span>{<span class="kw">&quot;</span><span class="st">gene1</span><span class="kw">&quot;</span>}){
    <span class="fu">print</span> <span class="kw">&quot;</span><span class="st">The key &#39;gene1&#39; exists in \%gene_info</span><span class="ch">\n</span><span class="kw">&quot;</span>;
}</code></pre>
<pre><code>## The key &#39;gene1&#39; exists in %gene_info</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="first-perl-program.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="control-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/OmicsAcademy/Novice2Expert4Bioinformatics/edit/master/11_PerlVariableOperator.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["bioinfBookXIE186.pdf", "bioinfBookXIE186.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
